enterprise/cmd/executor/internal/worker/workspace/files.go

```diff
@@ -73,12 +73,41 @@ type workspaceFile struct {
 	modifiedAt time.Time
 }
 
-var scriptPreamble = `
+// ScriptPreamble contains a script that checks at runtime if bash is available.
+// If it is, we want to be using bash, to support a more natural scripting.
+// If not, then we just run with sh.
+// This works roughly like the following:
+// - If no argument to the script is provided, this is the first run of it. We will use that later to prevent an infinite loop.
+// - Determine if a program called bash is on the path
+// - If so, we invoke this exact script again, but with the bash on the path, and pass an argument so that this check doesn't happen again.
+// - If not, it might be that PATH is not set correctly, but bash is still available at /bin/bash. If that's the case we do the same as above.
+// Otherwise we just continue and best effort run the script in sh.
+var ScriptPreamble = `
+# Only on the first run, check if we can upgrade to bash.
+if [ -z "$1" ]; then
+  bash_path=$(which bash)
```

Comment:
Today's installment of "Adam knows things he wishes he didn't": `which` isn't guaranteed to be available. (For example, it's not in the base `fedora` image.)

Instead, how about:

```suggestion
  bash_path=$(command -p -v bash)
```

Reply:
For the unenlightened, can you explain what this command does? As you might imagine, it's difficult to google effectively.

Reply:
`which` shows the full path of the first binary that is found in the `$PATH` that matches the first parameter. So, for example:

```
~ ❯ which bash
/opt/homebrew/anaconda3/bin/bash
```

[Here's the man page, if you're far too interested.](https://linux.die.net/man/1/which)

Reply:
We clarified in meatspace that the actual question was around `command`. Sorry.

`command` is [a POSIX-specified utility](https://pubs.opengroup.org/onlinepubs/9699919799/) — usually provided as a shell builtin — that provides ways of introspecting into commands available on the path.

Specifically: `-p` says to use the default value of `$PATH`, and `-v` says to output a "description" of the command, which for actual binaries will just be the absolute path name of the binary.